### ISSUE TITLE
Implement file IO using pure Inko

### ISSUE THREAD

**Issue by yorickpeterse** [inko-lang/inko#633]
### Description

With a proper FFI and cross compilation in place, there's no reason to keep the file IO functions implemented in the Rust runtime library (`inko_file_open`, `inko_file_write`, etc). This also gives the standard library the ability to operate on raw file descriptors where needed, making it easier to implement more file IO methods in the future.

The following functions need to be ported over:

- [x] inko_directory_create
- [x] inko_directory_create_recursive
- [x] inko_directory_remove
- [x] inko_directory_remove_recursive
- [x] inko_file_copy
- [x] inko_file_drop
- [x] inko_file_flush
- [x] inko_file_open
- [x] inko_file_read
- [x] inko_file_remove
- [x] inko_file_seek
- [x] inko_file_size
- [x] inko_file_write_bytes
- [x] inko_file_write_string
- [x] inko_path_accessed_at
- [x] inko_path_created_at
- [x] inko_path_exists
- [x] inko_path_expand
- [x] inko_path_is_directory
- [x] inko_path_is_file
- [x] inko_path_modified_at

### Blocked by

- https://github.com/inko-lang/inko/issues/583

**Comment by yorickpeterse** [comment]
I started work on this in the [inko-file-io](https://github.com/inko-lang/inko/tree/inko-file-io) branch. A few things I noted:

1. I forgot to rename a bunch of "Variant" instances to "Constructor", this is fixed
2. In f156a367f015bb4e0292058719f551139bda6d0c I forgot to take care of enum member fields, which are always defined as `Int`. This blows up the compiler and/or the program (depending on your luck) when storing values of a different size. We might have to introduce some sort of "blob" type that's just N bytes of space, then cast that to the appropriate type when loading the fields
3. There's a WIP setup for `ReadOnlyFile` that uses `open()`, `read()`, etc. This turned out to not be as tricky as I initially feared

**Comment by yorickpeterse** [comment]
Regarding item 2 from https://github.com/inko-lang/inko/issues/633#issuecomment-2305969299: `match` handles things just fine, since it operates on the final (specialized) types. Instead, it seems the drop logic doesn't handle this?

**Comment by yorickpeterse** [comment]
A simple way to reproduce this issue is this:

```inko
class enum Example {
  case A(Int32)
  case B(Int)
}

class async Main {
  fn async main {
    Example.A(42 as Int32)
  }
}
```

Running this results in a segfault at runtime.

**Comment by yorickpeterse** [comment]
I thought this would be the culprit:

https://github.com/inko-lang/inko/blob/988b72191f4aa66a3ac4f95dcf07b88634bad7de/compiler/src/mir/passes.rs#L560-L567

But changing this code such that it prints the register types shows it observes them to be correct.

**Comment by yorickpeterse** [comment]
The problem is the LLVM layout that we generate: it just uses the raw `Int` type as the field types. In addition, it seems we process the same class multiple times which doesn't seem quite right either.

**Comment by yorickpeterse** [comment]
Reading material: https://mapping-high-level-constructs-to-llvm-ir.readthedocs.io/en/latest/basic-constructs/unions.html

**Comment by yorickpeterse** [comment]
LLVM layouts for enums are now fixed, which brings us to the next problem: we seem to generate specialized methods for generic types that contain C types, which may then result in us generating a method that performs dynamic dispatch on such a type. That is, given a `Result[Int8, String]`, we might try to generate a `fmt` for that type.

The likely fix here is to consult the method bounds (if any), and only generate the method if the instance is compatible with it.

**Comment by yorickpeterse** [comment]
_Not_ generating methods for when the type arguments are _not_ compatible with bounds is a bit more tricky than expected. For one, we have to check _shapes_ not types since we specialize by shape and thus group multiple types into the same shape. This is doable, but it doesn't cover all cases. Most notably, there's a case where a previously dynamically dispatched call to `==` or `!=` on some generic type is turned into a statically dispatched call on `Result[Int32, String]`. This then results in the compiler trying to generate dynamic dispatch code on an `Int32`.

The core issue here is that type specialization operates on the assumption that all types support dynamic dispatch. Prior to f156a367f015bb4e0292058719f551139bda6d0c this was true, but since then this is no longer the case. There are two paths we can take here:

1. We detect the cases where we try to generate specialized methods for types that don't need/support them, then skip that.
2. We turn `Int8`, `Int16`, etc into first-class types and handle them like a regular `Int`, though perhaps with better names (e.g. `I8`, `U8`, etc instead of `Int8` and `UInt8`). This would at some point also allow us to remove `ByteArray` in favour of `Array[Uint8]`/`Array[U8]`

Adding support for different integer sizes _does_ require that we also add support for different integer literals (e.g. `10u8` or `10i32`), along with potentially other required changes. I'm a tad worried this results in a cascade of changes that may not be necessary.

**Comment by yorickpeterse** [comment]
Worth reading regarding signed vs unsigned integers: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1428r0.pdf

**Comment by yorickpeterse** [comment]
Regarding integer sizes: while a `Byte` (= unsigned int8) type makes sense, allowing us to replace `ByteArray` with `Array[Byte]`, types such as `Int16`, `Uint16`, `Int32`, and others just don't make that much sense. These are typically used when dealing with some serialization format or when trying to squeeze out every bit of memory, but neither is really relevant:

1. For serialization you can still read these kind of values and convert them to an `Int`
2. Inko isn't the kind of language you'd use in an environment where you have little memory (i.e. embedded), so there's simply less of a need for using int32 instead of int64.

You also end up with a never ending discussion of "should I use signed or unsigned integers?" Unsigned integers may be perceived as useful for e.g. indexing, but I'm not convinced the result is in fact better. For example, if the default integer type is `Int64` but indexing requires e.g. `Uint64`, users are just going to blindly do `the_index as Uint64`. In contrast, the current approach of using `Int` results in methods explicitly checking for negative values and producing useful error messages, instead of just silently wrapping the value.

**Comment by yorickpeterse** [comment]
Regarding specialization: it seems that when a trait provides a default method, and the trait is implemented with a set of bounds, the bounds aren't applied to the default method. This then prevents us from not processing the method when foreign types are assigned to the bounds.

**Comment by yorickpeterse** [comment]
This work revealed several compiler bugs, such as constant type inference and value evaluation/folding not working properly depending on the order in which they're processed. These have been resolved.

`ReadOnlyFile` no longer relies on any Rust code. The `Size` trait is in the process of being removed in favour of a `metadata` method for `Path` and the various file types. `ReadOnlyFile.metadata` is already implemented for Linux, using `statx()` to get the necessary data. 

The returned `Metadata` type handles the creation/birth time not always being available by returning a `Option[DateTime]`, while access and modification times are always present and are returned as a `DateTime`. I spent some time digging to see if there's ever a way for getting these values to fail, but this doesn't seem to be the case; at least not for the platforms we're interested in.

The next steps are:

1. Implement `Path.metadata` on Linux, which differs from `ReadOnlyFile.metadata` in that it doesn't have a file descriptor but a file path
2. Implement this logic for macOS and FreeBSD
4. Apply this to `WriteOnlyFile`, `ReadWriteFile`, and `Path`
5. Remove the `Size` trait
